Add hardware metadata files, generate setup scripts and hardware docs from them. (#883)
zmkfirmware/zmk: docs/docs/development/hardware-metadata-files.md

## File Format

The first line of every metadata file should contain the file format. As of today, the only acceptable file format is `1`, which should be written like:

```yaml
file_format: "1"
```

:::note

ZMK plans to expand on the initial based set of metadata properties, while maintaining backwards compatibility. In particular, new _optional_ properties may be added to file format `1` as time progress, and when new set of properties is settled upon as being required moving forward, a new major version of the file format will be created to encompass those changes.

:::

### Item ID and Name

All metadata files should then contain two key pieces of identifying information, the `id` for the item, and the `name` which used a the human readable display name in various UI locations:

```yaml
id: corne
name: Corne
```

### Item Types

Each metadata file includes a `type` property uniquely identifying the type of item, be it `board`, `shield`, or the less frequently needed `interconnect` (which is used to document generic hardware interconnects like the Pro Micro format):

```yaml
type: shield
```

### URL

> The `url` property should contain the canonical URL used to learn more about the board/shield, e.g. the main vendor website, or GitHub repository for a given keyboard/controller:

Reviewer: ```suggestion
The `url` property should contain the canonical URL used to learn more about the board/shield, e.g. the main vendor website or GitHub repository for a given keyboard/controller:
```

```yaml
url: https://github.com/foostan/crkbd/
```

### Interconnect Requires/Exposes

For boards and shields, one of the key pieces of high level information is compatibility between the two items. In particular, a board usually exposes one ore more "interconnects", the physical location/type of connections available, and their assigned possible uses (e.g. GPIO, power, ground, i2c, etc). Similarly, a shield is usually designed around one (or sometimes more) "interconnects" that allow it to connect to one of those boards.

> In ZMK, we encode both of those scenarios with the `exposes` and `requires` properties, respectfully. For example, for a Corne shield that requires a Pro Micro compatible controller to function, and simultaneously exposes a four pin header to be used by standard i2c OLED modules, the metadata file contains:

Reviewer: ```suggestion
In ZMK, we encode both of those scenarios with the `exposes` and `requires` properties, respectively. For example, for a Corne shield that requires a Pro Micro compatible controller to function and simultaneously exposes a four pin header to be used by standard i2c OLED modules, the metadata file contains:
```

Author: Thanks.

```yaml
requires:
  - pro_micro
exposes:
  - i2c_oled
```

### Features

Boards and shields should document the sets of hardware features found on them using the `features` array. There is a fixed enum of possible values to use here, which will be expanded over time. The current set of possible `features` values is:

> - `keys` - Any board or shield that contains keyboard keys should include this feature. It is a central feature used to determine if we have a "complete combination" for ZMK to produce a keyboard firmware when performig setup.

Reviewer 2: ```suggestion
- `keys` - Any board or shield that contains keyboard keys should include this feature. It is a central feature used to determine if we have a "complete combination" for ZMK to produce a keyboard firmware when performing setup.
```

- `display` - Indicates the hardware includes a display for use with the ZMK display functionality.
- `encoder` - Indicates the hardware contains one or more (possibly optional) rotary encoders.

> - `underglow` - Indicatges the hardware includes underglow LEDs.

Reviewer: ```suggestion
- `underglow` - Indicates the hardware includes underglow LEDs.
```

- `pointer` (future) - Used to indicate the hardware includes one or more pointer inputs, e.g. joystick, touchpad, or trackpoint.

### Siblings

> The `siblings` array is used to identify multiple hardware items designedc to be used together as one logical device. Right now, that primarily is used to identify the two halves of a split keyboard, but future enhancement will include more complicated and flexible combinations.

Reviewer: ```suggestion
The `siblings` array is used to identify multiple hardware items designed to be used together as one logical device. Right now, that primarily is used to identify the two halves of a split keyboard, but future enhancements will include more complicated and flexible combinations.
```

Author: Good catches. Fixed.
